text/0039-service-overhaul.md

```diff
@@ -0,0 +1,387 @@
+
+**TODO: Some user might have a removed service (e.g., Mixer)**
+
+**Downgrade might break service integration**
+Auth integration config inside `basic.ini` will be also transfered under a new section if in service settings.
+
+### Integrations
+
+Twitch, YouTube and Restream have their own integration in OBS with OAuth. But those need to be isolated as plugins.
+
+Those will plugins will provide the basic and OAuth version of the services as one services.
+
+If OBS is built without client ids and hashes, those plugins will be built without their integration.
+
+#### Dock state
+
+Actually the dock state is global in the config, but with docks that appear depending on the service which is per profile which is still "also" the case with "old" integrations.
+
+This can lead to dock state losses between profile switching and exiting. The "old" integration actually store a dock state in the profile config and restore it after the integration is is loaded.
+
+But making restoring a dock state from a plugin should should not be considered at all.
+Making it per profile could avoid this.
+
+Integrations docks will be added after specific frontend event.
+So after each of them the dock state needs to be restored.
+
+- The initial restoration of the profile dock state will be moved just after `OBS_FRONTEND_EVENT_FINISHED_LOADING`. 
+
+- `OBS_FRONTEND_EVENT_PROFILE_CHANGED`, profile dock state will be loaded after a profile is changes.
+
+Also they will be removed after specific frontend event.
+So before each of them the dock state needs to be saved.
+
+- `OBS_FRONTEND_EVENT_PROFILE_CHANGING`, profile dock state will be saved before a profile is changed.
+
+- `OBS_FRONTEND_EVENT_EXIT`, profile dock state must be saved before this event while OBS is exiting.
+
+#### Dock addition (**The concept needs to be tested**)
+
+Integration plugins will need events/signals to be able to add and remove their docks at the right moment.
+
+
+3 signals will be added to Core OBS signaling:
+
+- "service_create (ptr service)"
+- "service_update (ptr service)"
+- "service_destroy (ptr service)"
+
+A variant of `obs_service_create()` to spawn signal-less (and maybe even private) service will be needed to allow protocol checking in settings.
+
+The plugin will have to store if `OBS_FRONTEND_EVENT_FINISHED_LOADING` was passed or not.
+
+If created before `OBS_FRONTEND_EVENT_FINISHED_LOADING`, a front-event callback will be added to react to to later event to add docks (if integration connected).
+
+When created/updated (after `OBS_FRONTEND_EVENT_FINISHED_LOADING`) the service will add the docks (if integration connected).
+
+When `OBS_FRONTEND_EVENT_PROFILE_CHANGING` `OBS_FRONTEND_EVENT_EXIT` or the service is destroyed, docks will be removed (if integration connected).
+
+#### Browser features
+The Front-end API needs to enable the possibility to access some `obs-browser` related feature like adding browser docks and generating widgets.
```

Comment:
> Do services really need it ?

Except if you want to drop Twitch YouTube and Restream docks or create non-browser based alternative of those docks.
Otherwise yes those feature are needed since we completely stop having integrations directly inside the UI but as plugins.

I will elaborate the text in the future.

Reply:
> create non-browser based alternative of those docks

For docks like chats, this would take a while to implement.